### PR TITLE
kubeadm refactor cmd init

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR refactors kubeadm init command in order to use the phase runner and start graduating phases progressively

**Which issue(s) this PR fixes** : 
Rif https://github.com/kubernetes/kubeadm/issues/1163

**Special notes for your reviewer**:
The PR leaves TODOs to be addressed in following PRs that will effectively move phases from `kubeadm alpha` to `kubeadm init`

/sig cluster-lifecycle
/kind documentation
/assign @neolit123 
/assign @timothysc 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
@chuckha 
@rosti 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/phases/workflow/phase.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/phases/workflow/phase.go]: Too many `Get` prefixes in the methods of this interface, but that may be just me.

File: cmd/kubeadm/app/cmd/phases/workflow/runner_test.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/phases/workflow/runner_test.go]: Should `w` be defined here? Just so it's not modified by a test case.

**Review by rosti — COMMENTED** [review]
Thanks for this @fabriziopandini .
I had a chance to take a quick glance over it. Will do another pass tomorrow, but so far it looks good.

**Review by neolit123 — COMMENTED** [review]
@fabriziopandini 
thanks, the implementation seems smart and covers the desired use case.

added some random comments and questions here and there.
ideally we want more eyes on this.

File: cmd/kubeadm/app/cmd/phases/workflow/phase.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/phase.go]: This was intentional.

I think that the main usage of phases will be trough `SimplePhase` type instantiation
```
var myPhase = SimplePhase{
      Name:  "myPhase",
      Short: "A phase of a kubeadm composable workflow...",
      Run: func(ctx interface{}) error {
            ....
}
``` 
So I preserved names without suffix or prelix like e.g. `Name` for that struct

Instead, I think that the `Phase` interfaces (where I was forced to add `Get` prefix for avoiding name conflicts with ˆˆˆ)  will be used only if we need custom phase types along the way.

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/runner.go]: Changed to `FilterPhases` and `SkipPhases`

File: cmd/kubeadm/app/cmd/phases/workflow/runner.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/runner.go]: Phase runner supports this feature because currently there are nested phases and I think we should reproduce it.
I agree that this feature should not be abused when implementing phases.

File: cmd/kubeadm/app/cmd/phases/workflow/simplephase.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/simplephase.go]: Since I see no trouble in having separated files, I do prefer having the interface and the class implementing that interface in separated files.

File: cmd/kubeadm/app/cmd/phases/workflow/phase.go
Comment (neolit123) [reply cmd/kubeadm/app/cmd/phases/workflow/phase.go]: a semantical problem is that the getters (e.g. GetName) are acting on fields that are already public (e.g. Name).
the idea of a getter would be to make a private field public. (name -> Name())

do we really need the interface to be public? perhaps we can make it private to solve the getter naming issue.
we can potentially rename SimplePhase -> Phase and Phase can be just a struct like cobra Command.
then new phase types can use composition.

https://play.golang.org/p/kI7xZWpH6Xv

something else to note is that we don't really need the interface if all phases (consumers) are just going to create SimplePhase objects, instead of inheriting from SimplePhase.

File: cmd/kubeadm/app/cmd/phases/workflow/runner_test.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/runner_test.go]: done

File: cmd/kubeadm/app/cmd/phases/workflow/phase.go
Comment (fabriziopandini) [reply cmd/kubeadm/app/cmd/phases/workflow/phase.go]: We can talk about semantics for weeks 😉, but honestly I don't see this as a issue.

IMO the most important things is to getting phase moving, and we can eventually iterate on this detail afterwards

> something else to note is that we don't really need the interface if all phases 

I'm not sure about this, because might be we decide to implement a custom case for generating certs (accepting cert name as input) instead of generating ~10 simple phases.
This is why I'm keeping an interface in the picture

**Review by timothysc — CHANGES_REQUESTED** [review]
Can you PR just the workflow issue as it's own PR without the integration.  Once we agree on that we can integrate.  Right now it's a bit conflating. 

/cc @rdodev

**Comment by fabriziopandini** [comment]
/hold
Working on a seperated PR as suggested

**Comment by fabriziopandini** [comment]
/hold cancel
@timothysc @neolit123 now this PR is rebased on top of https://github.com/kubernetes/kubernetes/pull/69684 and now ready for a final review

**Review by neolit123 — APPROVED** [review]
LGTM, thanks.
seems good to merge after CI passes, unless others have comments.

**Comment by rdodev** [comment]
LGTM as well.

File: cmd/kubeadm/app/cmd/init.go
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: please rename the ctx variables per our other discussion around confusion.

**Review by timothysc — COMMENTED** [review]
Minor comment then LGTM 

thx @fabriziopandini !

File: cmd/kubeadm/app/cmd/init.go
Comment (rosti) [reply cmd/kubeadm/app/cmd/init.go]: Plus, you can simplify this to `ctx := ...` only.

File: cmd/kubeadm/app/cmd/init.go
Comment (rosti) [inline cmd/kubeadm/app/cmd/init.go]: It will be more clear if this is reduced to `AddInitOtherFlags(cmd.PersistentFlags(), &options)`.

Comment (dixudx) [reply cmd/kubeadm/app/cmd/init.go]: Agree with @rosti

Comment (timothysc) [reply cmd/kubeadm/app/cmd/init.go]: We can follow up on another PR.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 

Thx @fabriziopandini

**Pull request merged**
